language_bindings/rust/rust_build/src/lib.rs

```diff
@@ -0,0 +1,360 @@
+        let gen = self.new_gen("".to_string());
+
+        let compilation_result = gen.compile();
+        println!("Gen Creation Status: {}", compilation_result.status);
+        assert!(compilation_result.status.success());
+
+        let execution_result = gen.make_runtime();
+        println!("runtime results: {:?}", execution_result);
+        assert!(execution_result.is_ok());
+
+        gen.rename_runtime();
+    }
+}
+
+impl Generator<'static> {
+    ///Make the generator executable using Halide GenGen and g++
+    pub fn compile(&self) -> Output {
+        let mut cmd_compile = Command::new("g++");
+        cmd_compile.args(["-std=c++17"]);
+
+        // Inlcude flags
+        cmd_compile.args(["-I", self.halide_path.join("include").to_str().unwrap()]);
+        cmd_compile.args(["-I", self.halide_path.join("tools").to_str().unwrap()]);
+
+        // Linker flag
+        cmd_compile.args(["-L", self.halide_path.join("lib").to_str().unwrap()]);
+
+        let temp = self
+            .halide_path
+            .join("tools")
+            .join("GenGen")
+            .with_extension("cpp");
+        if self.gen_name.is_empty() {
+            cmd_compile.args(["-g", temp.to_str().unwrap()]);
+        } else {
+            cmd_compile.args([
+                "-g",
+                self.gen_path.to_str().unwrap(),
+                temp.to_str().unwrap(),
+            ]);
+        }
+        // Output flag
+        cmd_compile.args(["-o", self.gen_exe.to_str().unwrap()]);
+
+        cmd_compile.args(self.gcc_flags.clone());
+        //compile.args(["-Wl,-rpath","-Wl,/home/rootbutcher2/CLionProjects/Halide-Rusts-tests/Halide/distrib/lib/"]);
+        cmd_compile.output().expect("Make generator failed")
+    }
+    /// Runs the previously made executable
+    ///
+    /// Panics if unable to find or run the executable
+    ///```ignore
+    ///     gen.run_gen()
+    ///```
+    pub fn run_gen(&self) -> Output {
+        //assert!(!self.gen_exe.is_none());
+        println!(
+            "cargo:rustc-link-search=native={}",
+            self.rs_output.to_str().unwrap()
+        );
+
+        let mut gen = Command::new(self.gen_exe.to_str().unwrap());
+        gen.args(["-g", self.gen_name.as_str()]);
+        gen.args(["-f", self.gen_name.as_str()]);
+
+        //Todo change to build dir
+        gen.args(["-o", self.rs_output.to_str().unwrap()]);
+        gen.env("LD_LIBRARY_PATH", self.halide_path.join("lib"));
+
+        //Todo add debug
+        if !self.debug {
+            gen.arg(self.target.as_str());
+        } else {
+            let mut temp = String::from(self.target.as_str());
+            temp.push_str("-debug");
+            gen.arg(temp);
+        }
+        gen.output().expect("failed to run")
+    }
+
+    /// Make the Halide runtime
+    ///
+    /// The Halide runtime is required for using Halide and contains the buffer_t and other useful functions
+    pub fn make_runtime(&self) -> Result<()> {
```

Comment:
Is the purpose of this crate to compile Halide code, or to compile the Halide implementation itself, or both? These are different tasks that should be decoupled. In particular, tooling to compile Halide code should support a pre-compiled external Halide implementation, installed in the environment as documented at https://github.com/halide/Halide/#getting-halide, without any special effort from the caller. Only if none are available should this infrastructure attempt to compile the Halide implementation itself, and it should do so transparently to the caller (i.e. without requiring a path to source code be passed in).

Reply:
> Only if none are available should this infrastructure attempt to compile the Halide implementation itself, and it should do so transparently to the caller (i.e. without requiring a path to source code be passed in).

I would sooner see an error in this case, rather than attempting to build Halide from scratch. Which LLVM to use? Does the user have 10+ minutes to wait for Halide to build?

Reply:
That's reasonable; e.g. the Rust `openssl` crate will panic in its build script if OpenSSL can't be found in the environment. It can also be useful to allow a custom search path to be passed in via an environment variable (e.g. `HALIDE_PATH` perhaps?).